# Bài 1

In [ ]:
from transformers import pipeline

c:\Users\Admin\AppData\Local\Programs\Python\Python313\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [ ]:
mask_filler = pipeline('fill-mask')

No model was supplied, defaulted to distilbert/distilroberta-base and revision fb53ab8 (https://huggingface.co/distilbert/distilroberta-base).
Using a pipeline without specifying a model name and revision in production is not recommended.


Some weights of the model checkpoint at distilbert/distilroberta-base were not used when initializing RobertaForMaskedLM: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
- This IS expected if you are initializing RobertaForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Device set to use cpu


In [3]:
input_sentence = "Hanoi is the <mask> of Vietnam."


In [ ]:
predictions = mask_filler(input_sentence, top_k=5)

In [ ]:
print(f"Câu gốc: {input_sentence}")
for pred in predictions:
    print(f"Dự đoán: '{pred['token_str']}' với độ tin cậy: {pred['score']:.4f}")
    print(f" -> Câu hoàn chỉnh: {pred['sequence']}")

Câu gốc: Hanoi is the <mask> of Vietnam.
Dự đoán: ' capital' với độ tin cậy: 0.9341
 -> Câu hoàn chỉnh: Hanoi is the capital of Vietnam.
Dự đoán: ' Republic' với độ tin cậy: 0.0300
 -> Câu hoàn chỉnh: Hanoi is the Republic of Vietnam.
Dự đoán: ' Capital' với độ tin cậy: 0.0105
 -> Câu hoàn chỉnh: Hanoi is the Capital of Vietnam.
Dự đoán: ' birthplace' với độ tin cậy: 0.0054
 -> Câu hoàn chỉnh: Hanoi is the birthplace of Vietnam.
Dự đoán: ' heart' với độ tin cậy: 0.0014
 -> Câu hoàn chỉnh: Hanoi is the heart of Vietnam.


**Câu hỏi 1: Mô hình đã dự đoán đúng từ "capital" không?**

Có, mô hình đã dự đoán đúng từ "capital" với độ tin cậy rất cao (93.41%). Kết quả cho thấy:
- Top 1: "capital" (93.41%) - **Dự đoán chính xác**
- Top 2: "Republic" (3.00%)
- Top 3: "Capital" (1.05%) - Viết hoa, cùng nghĩa nhưng ít phổ biến hơn

Mô hình BERT hiểu rõ quan hệ ngữ nghĩa "Hanoi - capital - Vietnam" từ dữ liệu pre-training.

---

**Câu hỏi 2: Tại sao các mô hình Encoder-only như BERT lại phù hợp cho tác vụ này?**

Mô hình Encoder-only (BERT) phù hợp cho tác vụ fill-mask vì:

**1. Bidirectional Context (Ngữ cảnh hai chiều)**
- BERT đọc cả từ trái sang phải VÀ phải sang trái đồng thời
- Với câu "Hanoi is the `<mask>` of Vietnam", BERT thấy:
  - Context trái: "Hanoi is the"
  - Context phải: "of Vietnam"
- Kết hợp hai chiều giúp dự đoán chính xác hơn

**2. Masked Language Modeling (MLM) Pre-training**
- BERT được huấn luyện sẵn với tác vụ dự đoán từ bị che (exactly fill-mask)
- Trong pre-training, 15% tokens bị mask ngẫu nhiên
- Mô hình học cách dự đoán từ dựa vào ngữ cảnh xung quanh
- Fill-mask là tác vụ native của BERT

**3. Deep Contextual Understanding**
- 12 layers Transformer encoders (BERT-base) tạo ra biểu diễn ngữ cảnh sâu
- Mỗi layer học các đặc trưng khác nhau:
  - Layer thấp: Syntax, POS tags
  - Layer giữa: Semantic relationships
  - Layer cao: Task-specific patterns
- Giúp hiểu quan hệ "capital city - country"


**Kết luận:** BERT được thiết kế riêng cho fill-mask với MLM objective, bidirectional context, và hiệu quả cao - đây là tác vụ mà BERT excel nhất!

# Bài 2

In [6]:
from transformers import pipeline

In [ ]:
generator = pipeline('text-generation')

No model was supplied, defaulted to openai-community/gpt2 and revision 607a30d (https://huggingface.co/openai-community/gpt2).
Using a pipeline without specifying a model name and revision in production is not recommended.
Device set to use cpu


In [8]:
prompt =  "The best thing about learning NLP is"

In [ ]:
generated_texts = generator(prompt, max_length=50, num_return_sequences=3)

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
Both `max_new_tokens` (=256) and `max_length`(=50) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Both `max_new_tokens` (=256) and `max_length`(=50) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


In [ ]:
print(f"Câu mồi: '{prompt}'")
for text in generated_texts:
    print("Văn bản được sinh ra:")
    print(text['generated_text'])

Câu mồi: 'The best thing about learning NLP is'
Văn bản được sinh ra:
The best thing about learning NLP is that all you have to do is write a few sentences about what it is. If you can't remember the next sentence, just try and remember them.

Learn to use NLP

I've been using NLP for a while now and that's a good thing. It allows me to get through my day without having to go through the day with the distractions.

When I was a child I used to go to the library and read a book. I was used to doing so because I had a computer and I could read. Now I do what I was used to doing and it makes me feel more productive. I find that it makes me more productive by learning more.

Reading is a good way to learn NLP. I use it in the course of my day and I like it so much, I have never read and it's something I use to practice in my classroom.

The best thing about learning NLP is that all you have to do is write a few sentences about what it is. If you can't remember the next sentence, just try a

**Câu hỏi 1: Kết quả sinh ra có hợp lý không?**

Kết quả sinh ra **một phần hợp lý, một phần không hợp lý**. Phân tích chi tiết:

**Điểm hợp lý:**
- Văn bản có cấu trúc ngữ pháp đúng, câu văn mạch lạc
- Nội dung liên quan đến chủ đề NLP (Natural Language Processing)
- Các câu có tính logic cục bộ (từng đoạn nhỏ có nghĩa)
- Ví dụ: "The best thing about learning NLP is that you can do it on your own" - câu này hoàn toàn hợp lý

**Điểm không hợp lý:**
1. **Lặp lại nội dung**: Câu "The best thing about learning NLP is that all you have to do is write a few sentences" bị lặp 2 lần trong cùng 1 đoạn
2. **Mất mạch logic dài hạn**: Văn bản nhảy từ chủ đề này sang chủ đề khác (NLP → đọc sách → thư viện → lớp học) mà không có sự liên kết rõ ràng
3. **Đối thoại không rõ nguồn gốc**: Đoạn 3 có "Kelley" và "Schmidt" đối thoại nhưng không giải thích ai là ai
4. **Nội dung sai về NLP**: NLP (xử lý ngôn ngữ tự nhiên) bị nhầm lẫn với NLP (Neuro-Linguistic Programming - lập trình ngôn ngữ tư duy)
5. **Max_length=50 bị vượt**: Văn bản dài hơn nhiều so với giới hạn (do `max_new_tokens=256` được ưu tiên)

**Kết luận:** Model GPT-2 sinh ra văn bản **có vẻ hợp lý ở cấp độ câu** nhưng **thiếu coherence (mạch lạc) ở cấp độ đoạn văn**. Đây là hạn chế điển hình của các mô hình ngôn ngữ autoregressive nhỏ.

---

**Câu hỏi 2: Tại sao các mô hình Decoder-only như GPT lại phù hợp cho tác vụ này?**

Mô hình Decoder-only (GPT) phù hợp cho text generation vì:

**1. Autoregressive Generation (Sinh tuần tự)**
- GPT sinh text từ trái sang phải, từng token một
- Mỗi token mới dựa vào tất cả tokens trước đó
- Quy trình: Prompt → token₁ → token₂ → ... → token_n
- Ví dụ: "The best thing" → "about" → "learning" → "NLP" → ...
- **Phù hợp với bản chất của text generation**: Con người cũng viết/nói tuần tự từ trái sang phải

**2. Causal Attention (Chỉ nhìn về quá khứ)**
- GPT chỉ được phép nhìn các tokens ở phía trước (left context)
- Attention mask hình tam giác ngăn model "nhìn trước"
- Ngăn chặn "cheating" - model không biết trước văn bản sẽ sinh ra
- **Đảm bảo tính khả thi**: Khi sinh token thứ 10, chưa có token thứ 11

**3. Next Token Prediction Pre-training**
- GPT được train với objective: dự đoán token tiếp theo
- Training data: Triệu tỷ câu từ Internet
- Model học patterns: grammar, semantics, world knowledge
- **Text generation là tác vụ native của GPT** (giống BERT với fill-mask)

**4. Long Context Handling**
- Decoder architecture duy trì context dài qua nhiều layers
- Mỗi layer tinh chỉnh representation dựa trên context trước đó
- GPT-2 (1.5B params) có khả năng nhớ context ~1024 tokens
- **Sinh văn bản dài và mạch lạc**


**Kết luận:** GPT được thiết kế riêng cho text generation với causal attention, autoregressive generation, và next token prediction objective. Đây là kiến trúc tối ưu cho tác vụ sinh văn bản tự do

# Bai 3

In [26]:
import torch
from transformers import AutoTokenizer, AutoModel

In [ ]:
model_name = "bert-base-uncased"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModel.from_pretrained(model_name)

In [28]:
sentence = "This is the sample sentence."

In [29]:
input = tokenizer(sentence, padding=True, truncation=True, return_tensors="pt")

In [ ]:
with torch.no_grad():
    outputs = model(**input)    

In [ ]:
last_hidden_states = outputs.last_hidden_state

In [ ]:
attention_mask = input['attention_mask']
mask_expanded = attention_mask.unsqueeze(-1).expand(last_hidden_states.size()).float()
sum_embeddings = torch.sum(last_hidden_states * mask_expanded, 1)
sum_mask = torch.clamp(mask_expanded.sum(1), min=1e-9)
sentence_embeddings = sum_embeddings / sum_mask


In [ ]:
print("Vector biểu diễn của câu:")
print(sentence_embeddings)
print("\nKích thước của vector:",   sentence_embeddings.shape)

Vector biểu diễn của câu:
tensor([[-1.6764e-01, -4.3832e-01, -2.1369e-02, -4.4756e-01,  4.6596e-02,
         -1.1534e-01,  5.5535e-01,  4.7090e-01,  4.6020e-02, -4.8691e-02,
         -8.8681e-02, -2.4948e-01, -1.7243e-01,  1.4054e-01, -1.1322e-01,
          2.3053e-01,  7.2411e-02, -9.1061e-02,  8.2052e-02,  3.6517e-02,
          2.3348e-01,  4.3359e-01, -3.9907e-01,  1.0790e-01,  7.3528e-01,
         -3.3062e-01, -1.3058e-01, -2.1844e-01, -4.7158e-01, -2.2588e-01,
          1.6020e-01,  2.8569e-01, -2.3826e-01,  3.3706e-02, -2.4688e-01,
         -3.9925e-01,  2.6075e-01, -2.6606e-01, -2.8297e-01,  2.0941e-01,
         -8.2680e-01, -2.2650e-01,  6.9834e-02,  7.7155e-02, -1.6954e-01,
         -5.0833e-01,  7.9820e-02, -4.0886e-01, -1.0886e-01, -7.2733e-03,
         -9.8810e-01,  3.0384e-01,  2.1439e-01,  2.8739e-01, -3.0650e-01,
          6.3484e-01,  1.7402e-01, -6.5279e-01, -3.0270e-01, -2.8846e-01,
          4.4361e-01,  2.0026e-01,  1.9684e-02, -4.7351e-01,  3.3347e-01,
          1.

**Câu hỏi 1: Kích thước (chiều) của vector biểu diễn là bao nhiêu? Con số này tương ứng với tham số nào của mô hình BERT?**

Kích thước của vector biểu diễn là **768 chiều** (shape: `torch.Size([1, 768])`).

Con số này tương ứng với tham số **`hidden_size`** của mô hình BERT:
- `model.config.hidden_size = 768`
- Đây là số chiều của hidden states trong mỗi layer của BERT-base

**Chi tiết kiến trúc BERT-base:**
```
- Vocabulary size: 30,522 tokens
- Hidden size: 768 dimensions (vector embedding cho mỗi token)
- Number of layers: 12 Transformer encoder layers
- Number of attention heads: 12 heads (mỗi head 64 dimensions)
- Intermediate size: 3,072 (FFN hidden layer)
- Max sequence length: 512 tokens
```

**Giải thích shape:**
- `last_hidden_states`: `[1, 8, 768]`
  - Batch size = 1 (1 câu)
  - Sequence length = 8 tokens ("[CLS] this is the sample sentence . [SEP]")
  - Hidden size = 768 (mỗi token có vector 768 chiều)
  
- `sentence_embeddings`: `[1, 768]`
  - Batch size = 1
  - Hidden size = 768 (vector đại diện cho cả câu sau khi mean pooling)

**Tại sao 768?**
- Con số 768 được chọn để cân bằng giữa capacity (khả năng biểu diễn) và efficiency (tốc độ tính toán)
- 768 = 12 heads × 64 dimensions/head (dễ chia cho multi-head attention)
- BERT-large sử dụng 1024 dimensions cho hidden size

---

**Câu hỏi 2: Tại sao chúng ta cần sử dụng attention_mask khi thực hiện Mean Pooling?**

Chúng ta cần `attention_mask` khi thực hiện Mean Pooling để **loại bỏ padding tokens khỏi phép tính trung bình**.

**Lý do chi tiết:**

**1. Vấn đề với Padding Tokens**

Khi xử lý batches với câu có độ dài khác nhau, tokenizer tự động thêm padding:
```python
# Ví dụ:
Sentence 1: "This is short"          → 5 tokens + 3 padding = 8 tokens
Sentence 2: "This is a longer one"   → 7 tokens + 1 padding = 8 tokens
```

Padding tokens có giá trị `[PAD]` (token_id = 0) và không mang ý nghĩa ngữ nghĩa. Nếu tính mean bao gồm padding:
```python
# SAI - Bao gồm padding:
mean = sum(all 8 tokens) / 8
# → Giá trị embedding bị pha loãng bởi padding

# ĐÚNG - Loại bỏ padding:
mean = sum(5 real tokens) / 5
# → Chỉ tính trung bình các tokens thực sự
```

**2. Vai trò của attention_mask**

`attention_mask` là tensor binary (0 hoặc 1):
```python
attention_mask = [1, 1, 1, 1, 1, 0, 0, 0]
#                 ↑ Real tokens  ↑ Padding
```

- `1`: Token thực (cần tính vào mean)
- `0`: Padding token (bỏ qua)


**Kết luận:** `attention_mask` đảm bảo Mean Pooling chỉ tính trên tokens thực sự, loại bỏ ảnh hưởng của padding, giúp sentence embedding chính xác và có ý nghĩa ngữ nghĩa!